File: script/R/05_correlation_pruning.R
# ==============================================================================
# Step A — Correlation-based pruning
# - Carica le feature LOGO-selected
# - Pruning greedy per importanza: rimuove feature con |ρ| > soglia
# ==============================================================================
# Richiede: install.packages(c("ggplot2","reshape2"))
# ==============================================================================

library(ggplot2)
library(reshape2)

setwd(normalizePath("~/Desktop/IREN"))

CORR_THRESH <- 0.90

# ── Carica feature e importanze ───────────────────────────────────────────────
lines_raw   <- readLines("logo_selected_features.txt")
logo_feats  <- lines_raw[!grepl("^#", lines_raw) & nchar(trimws(lines_raw)) > 0]

imp_df <- read.csv("logo_feature_importance.csv", sep=";", stringsAsFactors=FALSE)
imp_df <- imp_df[imp_df$Feature %in% logo_feats, ]
imp_df <- imp_df[order(-imp_df$MeanImportance), ]
features_ranked <- imp_df$Feature
cat(sprintf("Feature LOGO-selected in input: %d\n", length(features_ranked)))

# ── Carica training set ───────────────────────────────────────────────────────
df <- read.csv("TRAIN_FEATURES.csv", sep=";", stringsAsFactors=FALSE)
X  <- df[, features_ranked, drop=FALSE]
for (col in names(X))
    if (any(is.na(X[[col]]))) X[[col]][is.na(X[[col]])] <- median(X[[col]], na.rm=TRUE)

corr_matrix <- cor(X)

# ── Pruning greedy ────────────────────────────────────────────────────────────
kept    <- character(0)
removed <- character(0)
reason  <- list()

for (feat in features_ranked) {
    if (length(kept) == 0) { kept <- c(kept, feat); next }
    corrs    <- abs(corr_matrix[feat, kept])
    max_corr <- max(corrs)
    if (max_corr <= CORR_THRESH) {
        kept <- c(kept, feat)
    } else {
        partner <- kept[which.max(corrs)]
        removed <- c(removed, feat)
        reason[[feat]] <- c(partner, round(corr_matrix[feat, partner], 4))
    }
}

cat(sprintf("\nSoglia correlazione: |ρ| > %.2f\n", CORR_THRESH))
cat(sprintf("Feature mantenute: %d\n", length(kept)))
cat(sprintf("Feature rimosse:   %d\n", length(removed)))

# ── Report ────────────────────────────────────────────────────────────────────
cat(sprintf("\n%s\n  FEATURE MANTENUTE (%d)\n%s\n", strrep("=",55), length(kept), strrep("=",55)))
for (i in seq_along(kept)) {
    imp <- imp_df$MeanImportance[imp_df$Feature == kept[i]]
    cat(sprintf("  %2d. %-15s  imp=%.5f\n", i, kept[i], imp))
}

cat(sprintf("\n  FEATURE RIMOSSE (%d) — troppo correlate:\n", length(removed)))
for (feat in removed) {
    imp     <- imp_df$MeanImportance[imp_df$Feature == feat]
    partner <- reason[[feat]][1]
    rho     <- as.numeric(reason[[feat]][2])
    cat(sprintf("  %-15s  imp=%.5f  |ρ|=%.3f  con %s\n", feat, imp, abs(rho), partner))
}

# ── Plot heatmap feature mantenute ───────────────────────────────────────────
corr_kept <- corr_matrix[kept, kept]
corr_melt <- melt(corr_kept)
names(corr_melt) <- c("Var1","Var2","value")
# Triangolo inferiore
corr_melt$value[as.integer(corr_melt$Var1) < as.integer(corr_melt$Var2)] <- NA

p <- ggplot(corr_melt, aes(x=Var1, y=Var2, fill=value)) +
    geom_tile(color="white") +
    scale_fill_gradient2(low="#B2182B", mid="white", high="#2166AC",
                         midpoint=0, limits=c(-1,1), na.value="gray95",
                         name="ρ") +
    geom_text(aes(label=ifelse(!is.na(value), sprintf("%.2f",value), "")),
              size=3) +
    labs(title=sprintf("Correlazione — %d feature post-pruning (|ρ|≤%.2f)",
                       length(kept), CORR_THRESH),
         x=NULL, y=NULL) +
    theme_minimal(base_size=10) +
    theme(axis.text.x=element_text(angle=45, hjust=1))

ggsave("corr_matrix_kept.png", p,
       width=max(6, length(kept)*0.5), height=max(5, length(kept)*0.45), dpi=150)
cat("\n✓  corr_matrix_kept.png\n")

# ── Salva ─────────────────────────────────────────────────────────────────────
rows <- lapply(features_ranked, function(feat) {
    imp <- imp_df$MeanImportance[imp_df$Feature == feat]
    if (feat %in% kept) {
        data.frame(Feature=feat, Status="Kept", MeanImportance=imp,
                   RemovedDueTo="", MaxCorr=NA_real_, stringsAsFactors=FALSE)
    } else {
        data.frame(Feature=feat, Status="Removed", MeanImportance=imp,
                   RemovedDueTo=reason[[feat]][1],
                   MaxCorr=as.numeric(reason[[feat]][2]), stringsAsFactors=FALSE)
    }
})
report_df <- do.call(rbind, rows)
write.table(report_df, "corr_pruning_report.csv", sep=";", row.names=FALSE, quote=FALSE)

con <- file("corr_pruned_features.txt","w")
writeLines(sprintf("# Correlation pruning — soglia |ρ|=%.2f", CORR_THRESH), con)
writeLines(sprintf("# Input: %d  →  Output: %d feature", length(features_ranked), length(kept)), con)
writeLines("", con)
writeLines(kept, con)
close(con)

cat("✓  corr_pruning_report.csv\n")
cat("✓  corr_pruned_features.txt\n")

File: script/Python/05_correlation_pruning.py
"""
Step A — Correlation-based pruning
- Carica le 31 feature LOGO-selected
- Matrice di correlazione su TRAIN
- Pruning greedy: ordina per importanza LOGO-CV (desc),
  per ogni feature rimuove quelle successive con |ρ| > soglia
- Output: corr_matrix.png, corr_pruned_features.txt, corr_pruning_report.csv
"""

import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import seaborn as sns

# ── Parametri ──────────────────────────────────────────────────────────────────
TRAIN_FILE   = "TRAIN_FEATURES.csv"
LOGO_FILE    = "logo_selected_features.txt"
LOGO_IMP_FILE= "logo_feature_importance.csv"
OUT_FEATS    = "corr_pruned_features.txt"
OUT_REPORT   = "corr_pruning_report.csv"
CORR_THRESH  = 0.90   # |ρ| > soglia → ridondante
# ──────────────────────────────────────────────────────────────────────────────

# ── Carica feature LOGO-selected ──────────────────────────────────────────────
with open(LOGO_FILE) as f:
    logo_feats = [l.strip() for l in f if l.strip() and not l.startswith("#")]
print(f"Feature LOGO-selected in input: {len(logo_feats)}")

# ── Carica ranking importanza da LOGO-CV ──────────────────────────────────────
imp_df = pd.read_csv(LOGO_IMP_FILE, sep=";")
# Mantieni solo le feature LOGO-selected, ordinate per importanza decrescente
imp_df = imp_df[imp_df["Feature"].isin(logo_feats)].sort_values(
    "MeanImportance", ascending=False
).reset_index(drop=True)
features_ranked = imp_df["Feature"].tolist()

# ── Carica training set e calcola correlazione ────────────────────────────────
df    = pd.read_csv(TRAIN_FILE, sep=";")
X_raw = df[features_ranked].values
X     = SimpleImputer(strategy="median").fit_transform(X_raw)

corr_matrix = pd.DataFrame(
    np.corrcoef(X.T),
    index=features_ranked,
    columns=features_ranked
)

# ── Correlation pruning greedy ────────────────────────────────────────────────
# Scorre le feature in ordine di importanza (dalla più alla meno importante).
# Tiene una feature se non è troppo correlata (|ρ| > soglia) con nessuna
# di quelle già tenute. In questo modo si preservano sempre le più importanti.
kept    = []
removed = []
reason  = {}   # feature → (correlata con, valore ρ)

for feat in features_ranked:
    if not kept:
        kept.append(feat)
        continue
    corrs = corr_matrix.loc[feat, kept].abs()
    max_corr = corrs.max()
    if max_corr <= CORR_THRESH:
        kept.append(feat)
    else:
        corr_partner = corrs.idxmax()
        removed.append(feat)
        reason[feat] = (corr_partner, round(corr_matrix.loc[feat, corr_partner], 4))

print(f"\nSoglia correlazione: |ρ| > {CORR_THRESH}")
print(f"Feature mantenute: {len(kept)}")
print(f"Feature rimosse:   {len(removed)}")

# ── Report ────────────────────────────────────────────────────────────────────
print(f"\n{'':=<60}")
print(f"  FEATURE MANTENUTE ({len(kept)})")
print(f"{'':=<60}")
for i, feat in enumerate(kept):
    imp = imp_df[imp_df["Feature"]==feat]["MeanImportance"].values[0]
    print(f"  {i+1:>2}. {feat:<15}  imp={imp:.5f}")

print(f"\n  FEATURE RIMOSSE ({len(removed)}) — troppo correlate con una più importante:")
for feat in removed:
    partner, rho = reason[feat]
    imp = imp_df[imp_df["Feature"]==feat]["MeanImportance"].values[0]
    print(f"  {feat:<15}  imp={imp:.5f}  |ρ|={rho:.3f}  con {partner}")

# ── Plot matrice di correlazione (feature mantenute) ─────────────────────────
corr_kept = corr_matrix.loc[kept, kept]
n = len(kept)
fig, ax = plt.subplots(figsize=(max(8, n*0.5), max(7, n*0.45)))
mask = np.triu(np.ones_like(corr_kept, dtype=bool), k=1)
sns.heatmap(corr_kept, mask=mask, annot=n<=20, fmt=".2f",
            cmap="RdBu_r", vmin=-1, vmax=1, center=0,
            linewidths=0.4, ax=ax,
            annot_kws={"size": 7})
ax.set_title(f"Correlazione — {n} feature post-pruning (|ρ|≤{CORR_THRESH})", fontsize=12)
plt.tight_layout()
plt.savefig("corr_matrix_kept.png", dpi=150)
plt.close()
print(f"\n✓  corr_matrix_kept.png")

# Plot completo (tutte le 31, con evidenza delle rimosse)
fig2, ax2 = plt.subplots(figsize=(max(10, len(features_ranked)*0.45),
                                   max(9, len(features_ranked)*0.40)))
mask2 = np.triu(np.ones_like(corr_matrix, dtype=bool), k=1)
sns.heatmap(corr_matrix, mask=mask2,
            cmap="RdBu_r", vmin=-1, vmax=1, center=0,
            linewidths=0.3, ax=ax2, annot=False)
ax2.set_title(f"Correlazione — tutte le 31 feature LOGO-selected", fontsize=11)
plt.tight_layout()
plt.savefig("corr_matrix_full.png", dpi=150)
plt.close()
print(f"✓  corr_matrix_full.png")

# ── Salva report CSV ──────────────────────────────────────────────────────────
rows = []
for feat in features_ranked:
    imp = imp_df[imp_df["Feature"]==feat]["MeanImportance"].values[0]
    if feat in kept:
        rows.append({"Feature": feat, "Status": "Kept",
                     "MeanImportance": imp, "RemovedDueTo": "", "MaxCorr": ""})
    else:
        partner, rho = reason[feat]
        rows.append({"Feature": feat, "Status": "Removed",
                     "MeanImportance": imp, "RemovedDueTo": partner, "MaxCorr": rho})

report_df = pd.DataFrame(rows)
report_df.to_csv(OUT_REPORT, sep=";", index=False)
print(f"✓  {OUT_REPORT}")

# ── Salva lista feature mantenute ─────────────────────────────────────────────
with open(OUT_FEATS, "w") as f:
    f.write(f"# Correlation pruning — soglia |ρ|={CORR_THRESH}\n")
    f.write(f"# Input: {len(logo_feats)} feature  →  Output: {len(kept)} feature\n\n")
    for feat in kept:
        f.write(feat + "\n")
print(f"✓  {OUT_FEATS}")
